Add support for preemptive scheduling in simulator build

sebastianene07 opened the pull request:
## Summary

This patch adds support for preemptive scheduling in NuttX simulator build. It uses a host system timer to generate periodic signals which act like a hardware tick. A new defconfig has been added for the simulator build named ```nsh3```.

## Impact

The simulation would be more responsive using this approach, as we don't have to deliver the events from the ```Idle Task```. 
If a task is ending up in an infinite closed loop, it will be preempted as opposed to locking the entire simulation.

## Testing

To enable this feature : ``` CONFIG_SIM_PREEMPTIBLE ``` should be enabled from Kconfig.
The patch can be tested by using this config:

```
./tools/configure.sh -a ../incubator-nuttx-apps sim/nsh3
```

Testing the simulator build on OSX requires allocating ~128k of stack size for tasks.

Signed-off-by: Sebastian Ene <nuttx@fitbit.com>

patacongo commented:
@yamt @xiaoxiang781216 You two are the primary uses of the simulator these days.  Can you review, approve, and merged this PR?  I will stay out of the loop on this one.

liuguo09 commented:
> There are some coding style problems that need to be fixed.
> 
> Also, these files no long exist in the current repository:
> 
> ```
> sched/errno/errno_errno.c: fatal: Failed to open
> sched/errno/errno_get.c: fatal: Failed to open
> sched/errno/errno_set.c: fatal: Failed to open
> syscall/syscall_lookup.h: fatal: Failed to open
> ```

I think this is caused by my CI change. I have made PR https://github.com/apache/incubator-nuttx/pull/1013 to revert it now.

patacongo commented:
> > There are some coding style problems that need to be fixed.
> > Also, these files no long exist in the current repository:
> > ```
> > sched/errno/errno_errno.c: fatal: Failed to open
> > sched/errno/errno_get.c: fatal: Failed to open
> > sched/errno/errno_set.c: fatal: Failed to open
> > syscall/syscall_lookup.h: fatal: Failed to open
> > ```
> 
> I think this is caused by my CI change. I have made PR #1013 to revert it now.

I think it is because I removed those files in a PR today

xiaoxiang781216 commented:
@sebastianene07 thanks for patch, but sim for Linux and macOS hit some build issue, please take a look.

yamt commented:
does windows have ucontext stuff?
(just a question. i don't know the current state of windows support.)

yamt commented on arch/sim/include/irq.h:
```diff
@@ -44,6 +44,10 @@
  ****************************************************************************/
 
+#ifdef CONFIG_SIM_PREEMPTIBLE
+#  include <ucontext.h>
```
to avoid mixing nuttx api and host api, we usually don't use the host api directly.
we have a set of special files to interact with host api. (HOSTSRCS)

sebastianene07 replied:
I did this because I needed to keep the ```ucontext_t``` structure inside the TCB. Do you have a better alternative ?

xiaoxiang781216 replied:
Two method I can come up:
1.Define XCPTCONTEXT_REGS bigger than ucontext storage
2.Define a pointer and allocate memory dynamically
But, anyway the code build inside NuttX can't include host side header file. It is't acceptable from the techincal point.

sebastianene07 replied:
As I'm working on my new patch (to switch from ucontext API  to sigsetjmp/siglongjmp pair) I need to keep the sigjmp_buf inside the xcpcontext. The problem is that you can't simply use a pointer and use dynamic memory allocation because the heap memory is not initialized when up_initial_state() is called for the Idle Task. We can have an array that can hold the sigjmp_buf data but the size of this is arch dependent. Do you have a better solution to this ?

xiaoxiang781216 replied:
Two methods I can imagine:
1.Call malloc/free function on host side just like you call host timer API
2.Define an array of sigjmp_buf like this:
```
  sigjmp_buf g_jmpbuf[CONFIG_MAX_TASKS];
```
and access it through pid:
```
g_jmpbuf[PIDHASH(pid))]
```
BTW, you don't need save pointer inside xcpcontext.

sebastianene07 replied:
That makes sense, thanks !

xiaoxiang781216 replied:
BTW, you can't call NuttX malloc in up_initial_state because the critical section already enter.

sebastianene07 replied:
I used static allocation for this

xiaoxiang781216 commented on arch/sim/src/sim/up_blocktask.c:
```diff
@@ -118,7 +118,11 @@ void up_block_task(struct tcb_s *tcb, tstate_t task_state)
        * value, then this is really the previously running task restarting!
        */
 
+#ifdef CONFIG_SIM_PREEMPTIBLE
```
It's better to combine this patch with:
https://github.com/apache/incubator-nuttx/pull/1009/commits/0e2bee8c4731ffa616a48e270be346947a4b489b

xiaoxiang781216 commented on arch/sim/src/sim/up_initialstate.c:
```diff
@@ -73,6 +73,11 @@ void up_initial_state(struct tcb_s *tcb)
 #else
   int ret;
 
+  if (tcb->adj_stack_ptr == NULL)
```
Merge with the previous patch, we need each patch can work correctly.

xiaoxiang781216 commented on arch/sim/src/sim/up_idle.c:
```diff
@@ -89,6 +89,15 @@ void up_idle(void)
+   * Hopefully, something will wake up.
+   */
+
+  nxsched_process_timer();
```
Why CONFIG_SIM_PREEMPTIBLE can't work with CONFIG_SCHED_TICKLESS?

sebastianene07 replied:
We no longer needed the fake timer interrupts delivered from ```Idle Task``` because we can rely on the host timer to drive the scheduler. What do you think ?

xiaoxiang781216 replied:
Yes, but if you support tickless, you should deliver the timer interrupt according OS request. Now you generate the timer interrupt at a fixed interval. Here is more info you can reference:
https://cwiki.apache.org/confluence/display/NUTTX/NuttX+Simulation
That's why I suggest you don't add your own timer implementation, the right thing is enhancing up_oneshot.c which already support both fixed tick and variable tick.

xiaoxiang781216 replied:
> We no longer needed the fake timer interrupts delivered from `Idle Task` because we can rely on the host timer to drive the scheduler. What do you think ?

But we can reuse the most logic inside up_oneshot.c, add the alarm tmer logic and empty up_timer_update.

sebastianene07 replied:
I reused the logic from up_oneshot.c, thanks for the feedback

xiaoxiang781216 commented on arch/sim/src/sim/up_timer.c:
```diff
@@ -0,0 +1,169 @@
+#include <sys/time.h>
+#include <sys/wait.h>
+
+#include "../../../../include/nuttx/config.h"
```
Why need include this file?

sebastianene07 replied:
I wanted to have access to ``` CONFIG_SIM_PREEMPTIBLE_TICK_DURATION_MS ``` definition

xiaoxiang781216 replied:
Please use the following method to get CONFIG_xxx value from host side:
ifeq ($(CONFIG_SMP),y)
  CSRCS += up_smpsignal.c up_cpuidlestack.c
  REQUIREDOBJS += up_smpsignal$(OBJEXT)
  HOSTCFLAGS += -DCONFIG_SMP=1 -DCONFIG_SMP_NCPUS=$(CONFIG_SMP_NCPUS)
  HOSTCFLAGS += -DCONFIG_USEC_PER_TICK=$(CONFIG_USEC_PER_TICK)
ifeq ($(CONFIG_SIM_WALLTIME),y)
  HOSTCFLAGS += -DCONFIG_SIM_WALLTIME=1
endif
  HOSTSRCS += up_simsmp.c
  STDLIBS += -lpthread
endif

xiaoxiang781216 commented on arch/sim/src/sim/up_timer.c:
```diff
@@ -0,0 +1,169 @@
+/****************************************************************************
+ * arch/sim/src/sim/up_timer.c
+ *
+ *   Copyright (C) 2014, 2019 Gregory Nutt. All rights reserved.
```
All new file need Apache copyright, you can find the right head from the code base.

xiaoxiang781216 commented on arch/sim/src/sim/up_timer.c:
```diff
@@ -0,0 +1,169 @@
+ ****************************************************************************/
+static void host_signal_handler(int sig, siginfo_t *si, void *old_ucontext)
+{
+  g_sched_process_timer_cb();
```
Does this really work? The signal can happen at any point, I don't think nxsched_process_timer can be called without any protection.

sebastianene07 replied:
Most of the operations in ```nxsched_process_timer``` are done in a critical section. In our case ```up_irq_save``` calls ```host_signal_save``` which masks the signal in the current host thread.

xiaoxiang781216 replied:
Yes, I saw your patch which disable/enable signal in save/restore irq. That's why ask you reorg your patch set to ensure each patch work correctly. From my review, most patch need merge into one.

xiaoxiang781216 commented on arch/sim/src/sim/up_timer.c:
```diff
@@ -0,0 +1,169 @@
+ *
+ ****************************************************************************/
+
+int up_init_timer(void (*nxsched_process_timer)(void))
```
I think this logic is better integrate into up_oneshot.c. The timer interrupt generation is very easy, the hardest part is how to call nxsched_process_timer without synchronization issue.

xiaoxiang781216 commented on arch/sim/src/sim/up_initialstate.c:
```diff
@@ -73,6 +73,10 @@ void up_initial_state(struct tcb_s *tcb)
   int ret;
 
+#if CONFIG_RR_INTERVAL > 0
+  tcb->flags |= TCB_FLAG_SCHED_RR;
```
The change is wrong, CONFIG_RR_INTERVAL doesn't mean all thread will use the route robin policy.

sebastianene07 replied:
I saw that if threads don't have ```TCB_FLAG_SCHED_RR``` flag set we won't be able to execute ```sched_roundrobin_process``` in ``` nxsched_cpu_scheduler ``` file : sched/sched/sched_processtimer.c +84

xiaoxiang781216 replied:
Yes, because only the thread enable roundrobin policy, then we need to execute sched_roundrobin_process. There are other policies exist(e..g FIFO or SPORADIC).
https://linux.die.net/man/2/sched_setscheduler

xiaoxiang781216 commented on boards/sim/sim/sim/configs/nsh3/defconfig:
```diff
@@ -19,11 +19,26 @@ CONFIG_BOARD_LOOPSPERMSEC=0
 CONFIG_EXAMPLES_HELLO=y
 CONFIG_EXAMPLES_HELLO_STACKSIZE=8192
+CONFIG_EXAMPLES_TELNETD=y
+CONFIG_EXAMPLES_TELNETD_CLIENTSTACKSIZE=8192
```
you can change DEFAULT_TASK_STACKSIZE instead.
And please merge with the first nsh3/defconfig

sebastianene07 replied:
I no longer have to change this (the enormous stack size was used by the ucontext API)

xiaoxiang781216 commented on arch/sim/src/sim/up_interruptcontext.c:
```diff
@@ -43,10 +43,47 @@
 #include "up_internal.h"
 
+#ifdef CONFIG_SIM_PREEMPTIBLE
+uint32_t host_signal_save(void);
```
@sebastianene07 please merge all preemptive scheduling related change in one patch. Without this patch, your previous change just break the system.

sebastianene07 replied:
Ok I'll squash all these commits in a single one. Is that ok ?

xiaoxiang781216 replied:
Maybe like this:
1.The alarm timer
2.ucontext
3.nsh3 config
But I suggest that:
1.Move alarm timer into up_oneshot.c We don't need create another timer implementation, it's better to enhance up_oneshot.c directly which is much more simple. I suppose you just need add 100~ lines to support this feature.
2.We don't need save/restore context by ucontext API, so it's better to drop ucontext patch.

sebastianene07 replied:
I removed the ucontext API

xiaoxiang781216 commented:
@sebastianene07 I stop to review the last several patch because it's better to reorg your patch into a meaning set. For example, I found serveral patch fix the problem in the previous one.
And basically, the better change is enhance up_oneshot.c which is much simple and get the tickless support free:
1.Register alarm timer in up_oneshot.c
2.Call back to NuttX in signal action
3.Modify up_irq_save/up_irq_restore to disable/enable signal
I don't think it's nessary to use ucontext API.

patacongo commented:
> 
> 
> does windows have ucontext stuff?
> (just a question. i don't know the current state of windows support.)

Windows has several different environments, WSL, Cygwin, MSYS, and windows native.  WSL should not be an issue.  Niether should Cygwin -- but we need to verify that.  MSYS uses the Cygwin libraries and should be okay.

AFAIK no one has ever attempted the simulator in a Windows native environment.  Certain POSIX libraries would be needed in that case.  Microsoft recommends:   https://docs.microsoft.com/en-us/cpp/porting/porting-from-unix-to-win32?view=vs-2019

The MinGW build would be a "native" build. MinGW does support ucontext.h, but I do not know the state: https://github.com/rantoniello/mingw-w64_selfcontained/blob/master/gcc-8.1.0/fixincludes/tests/base/sys/ucontext.h 

There is also this:  https://www.codeproject.com/Tips/4225/Unix-ucontext-t-Operations-on-Windows-Platforms

Thin MinGW/Native build would still most likely be challenging.

sebastianene07 commented:
> @sebastianene07 I stop to review the last several patch because it's better to reorg your patch into a meaning set. For example, I found serveral patch fix the problem in the previous one.
> And basically, the better change is enhance up_oneshot.c which is much simple and get the tickless support free:
> 1.Register alarm timer in up_oneshot.c
> 2.Call back to NuttX in signal action
> 3.Modify up_irq_save/up_irq_restore to disable/enable signal
> I don't think it's nessary to use ucontext API.

I used ucontext API because it can restore the signal mask after a call to ```swapcontext```. The setjmp/longjmp mechanism doesn't seem to work with signals. Do you think it's better to close this PR and re-open a different one with the suggested changes ?

xiaoxiang781216 commented:
> > @sebastianene07 I stop to review the last several patch because it's better to reorg your patch into a meaning set. For example, I found serveral patch fix the problem in the previous one.
> > And basically, the better change is enhance up_oneshot.c which is much simple and get the tickless support free:
> > 1.Register alarm timer in up_oneshot.c
> > 2.Call back to NuttX in signal action
> > 3.Modify up_irq_save/up_irq_restore to disable/enable signal
> > I don't think it's nessary to use ucontext API.
> 
> I used ucontext API because it can restore the signal mask after a call to `swapcontext`. The setjmp/longjmp mechanism doesn't seem to work with signals. Do you think it's better to close this PR and re-open a different one with the suggested changes ?

But if the irq save/restore work correctly, I think we don't need care about the host signal logic. The thread in NuttX isn't 1:1 mapping to host thread. The whole NuttX simulator run on one host thread(main thread) except you enable SMP then we have multiple threads(same number as Simulated CPUs). So the context switch inside NuttX doesn't mean we switch host thread. From the host view, the thread is still running on the main thread.

sebastianene07 commented:
> > > @sebastianene07 I stop to review the last several patch because it's better to reorg your patch into a meaning set. For example, I found serveral patch fix the problem in the previous one.
> > > And basically, the better change is enhance up_oneshot.c which is much simple and get the tickless support free:
> > > 1.Register alarm timer in up_oneshot.c
> > > 2.Call back to NuttX in signal action
> > > 3.Modify up_irq_save/up_irq_restore to disable/enable signal
> > > I don't think it's nessary to use ucontext API.
> > 
> > 
> > I used ucontext API because it can restore the signal mask after a call to `swapcontext`. The setjmp/longjmp mechanism doesn't seem to work with signals. Do you think it's better to close this PR and re-open a different one with the suggested changes ?
> 
> But if the irq save/restore work correctly, I think we don't need care about the host signal logic. The thread in NuttX isn't 1:1 mapping to host thread. The whole NuttX simulator run on one host thread(main thread) except you enable SMP then we have multiple threads(same number as Simulated CPUs). The context switch inside NuttX doesn't mean we switch host thread. From the host view, the thread is still running on the main thread, so I can't understand why we need save/restore host signal flag here.

We need to care about the host signal logic because when we are executing the timer signal handler we usually end up altering the signal handler execution flow(by doing a longjmp). Altering the execution flow of the signal handler without restoring the signal mask may lead to no further events being served.

xiaoxiang781216 commented:
Ok, I see. ucontext is required to jump inside signal handler, so it's better to have:
1.one patch implement the premptive context switch completely
2.one patch add a new config to demo this feature.

xiaoxiang781216 commented on arch/sim/src/Makefile:
```diff
@@ -71,10 +71,16 @@ CSRCS += up_unblocktask.c up_blocktask.c up_releasepending.c
 DEPPATH = $(patsubst %,--dep-path %,$(subst :, ,$(VPATH)))
 
 HOSTSRCS = up_hosttime.c
+HOSTCFLAGS += -DCONFIG_SIM_PREEMPTIBLE_TICK_DURATION_MS=$(CONFIG_SIM_PREEMPTIBLE_TICK_DURATION_MS)
```
I suggested to use USEC_PER_TICK before.

sebastianene07 replied:
I will update my patch

xiaoxiang781216 commented on arch/sim/src/sim/up_oneshot.c:
```diff
@@ -357,6 +358,16 @@ FAR struct oneshot_lowerhalf_s *oneshot_initialize(int chan,
+   */
+
+#ifdef CONFIG_SIM_PREEMPTIBLE
+  up_prepare_timer();
```
As I said before, timer isn't compatible with oneshot timer, both can't work together. you should integrate the alarm timer logic into oneshot.

sebastianene07 replied:
Thanks for the suggestion, ```up_oneshot.c``` is part of the CSRCS list and it's using the NuttX API. My timer is using the host timer and it builds in the HOSTSRC list. I am not sure if merging them it's ok. What do you think ?

xiaoxiang781216 replied:
Yes, up_oneshot.c is the bridge between NuttX and host. up_oneshot.c implmeent the standard NuttX timer driver interface and forward the real timing call(host_sleepuntil) to up_hosttime.c which is part of HOSTSRC. Here is the basic idea to integrate your code into up_oneshot.c:
1.Implment the alaram wrapper and signal handler in up_hosttime.c
2.These functions should forward to your alarm wrapper in up_hosttime.c:
  oneshot_initialize
  sim_start
  sim_cancel
  it's almost 1:1 mapping for alarm API
3.Signal handler should call to some function(similar with up_timer_update) in up_hosttime.c
Of course, we should make sure the code can run without problem with/without CONFIG_SIM_PREEMPTIBLE.

xiaoxiang781216 commented on boards/sim/sim/sim/scripts/Make.defs:
```diff
@@ -156,3 +158,4 @@ HOSTCC = cc
 HOSTINCLUDES = -I.
 HOSTCFLAGS = $(ARCHWARNINGS) $(ARCHOPTIMIZATION) \
    $(ARCHCPUFLAGS) $(HOSTINCLUDES) $(ARCHDEFINES) $(EXTRAFLAGS) -pipe
+HOSTCFLAGS += -DCONFIG_USEC_PER_TICK=$(CONFIG_USEC_PER_TICK)
```
Move to arch/sim/sim/src/Makefile

sebastianene07 replied:
Ok will do 👍

xiaoxiang781216 commented on include/signal.h:
```diff
@@ -267,9 +267,9 @@
  * special meaning in some circumstances (e.g., kill()).
  */
 
-#ifndef __SIGSET_T_DEFINED
```
After some thought, your modify this just because you include ucontext.h directly? if so, it's better to don't bring ucontext.h into NuttX space. then we don't need modify this file anymore.

sebastianene07 replied:
Right, this is why I needed this change

xiaoxiang781216 replied:
That's why it's better to not mix the NuttX and host header file which introduce many headache. And only allow source code in HOSTSRC can include the host header file.

xiaoxiang781216 commented on arch/sim/src/sim/up_unblocktask.c:
```diff
@@ -129,7 +133,11 @@ void up_unblock_task(FAR struct tcb_s *tcb)
           /* Then switch contexts */
 
+#ifdef CONFIG_SIM_PREEMPTIBLE
+          swapcontext(&prev_rtcb->xcp.context, &rtcb->xcp.context);
```
As I said before, you shouldn't call host function directly inside NuttX code.

sebastianene07 replied:
I removed this call

xiaoxiang781216 commented on arch/sim/src/sim/up_timer.c:
```diff
@@ -0,0 +1,362 @@
+ *
+ ****************************************************************************/
+
+int up_prepare_timer(void)
```
Please merge all related logic into up_oneshot.c. It doesn't make sense that we implment timer twice.

sebastianene07 replied:
I merged all the logic in up_oneshot.c, thanks for the feedback

xiaoxiang781216 commented on arch/sim/src/sim/up_releasepending.c:
```diff
@@ -114,7 +118,11 @@ void up_release_pending(void)
           /* Then switch contexts */
 
+#ifdef CONFIG_SIM_PREEMPTIBLE
+          swapcontext(&prev_rtcb->xcp.context, &rtcb->xcp.context);
```
Why not use sigsetjmp and siglongjmp:
https://linux.die.net/man/3/sigsetjmp

sebastianene07 replied:
By looking more at this API `sigsetjmp/siglongjmp` doesn't have a direct way for setting the initial task context. The ucontext API has the advantage of having the makecontext function. We need to manually change the PC and the SP in the `sigjmp_buf` from the `up_initial_state` call and this should be done for every arch.

xiaoxiang781216 commented:
> I used ucontext API because it can restore the signal mask after a call to `swapcontext`. The setjmp/longjmp mechanism doesn't seem to work with signals. Do you think it's better to close this PR and re-open a different one with the suggested changes ?

After reading the related document carefully, I think at least there are at least two other possible, simple and portable method than ucontext:
1.call sigaction with SA_NODEFER
http://man7.org/linux/man-pages/man2/sigaction.2.html 
2.call sigsetjmp and siglongjmp
https://linux.die.net/man/3/sigsetjmp

sebastianene07 commented:
@xiaoxiang781216 We still need a way to create and switch to a new stack so ```ucontext``` functions can't be replaced with the sigsetjmp/siglongjmp. I looked at the Qemu implementation and it seems that they are using ucontext for task setup & creation and sigsetjmp/siglongjmp for everything else.

xiaoxiang781216 commented:
Yes, it is hard to find a portable way to change SP/PC in jmpbuf.

xiaoxiang781216 commented:
Yes, it is hard to find a portable way to change SP/PC in jmpbuf.

xiaoxiang781216 commented:
Yes, the SP modification need adjust for each host, @sebastianene07 can reference state-thread which implement user mode thread by setjmp/longjmp:
https://github.com/ossrs/state-threads/blob/srs/md.h#L317
But there is a portable way to modify PC like this:
```
void up_initial_state(struct tcb_s *tcb)
{
  if (tcb->pid == 0)
    {
      tcb->adj_stack_size = CONFIG_IDLETHREAD_STACKSIZE;
      tcb->stack_alloc_ptr = (void *)((uintptr_t)host_get_topstackaddr() - CONFIG_IDLETHREAD_STACKSIZE);
    }

  if (setjmp((&tcb->xcp))
    {
       tcb->start();
    }
  else
    {
       tcb->xcp.regs[JB_SP] = (xcpt_reg_t)tcb->adj_stack_ptr - sizeof(xcpt_reg_t);
    }
}
```

sebastianene07 commented on include/signal.h:
```diff
@@ -267,8 +267,9 @@
  */
 
-#ifndef __SIGSET_T_DEFINED
+#if !defined(__sigset_t_defined) && !defined(__SIGSET_T_DEFINED)
```
I removed the changes from this header file

sebastianene07 commented:
> SP modification need adjust for each host, @sebastianene07 can reference state-thread which implement user mode thread by setjmp/longjmp:
> https://github.com/ossrs/state-threads/blob/srs/md.h#L317
> But there is a portable way to modify PC like this:
> 
> ```
> void up_initial_state(struct tcb_s *tcb)
> {
>   if (setjmp((&tcb->xcp))
>     {
>        tcb->start();
>     }
>   else
>     {
>        tcb->xcp.regs[JB_SP] = (xcpt_reg_t)tcb->adj_stack_ptr - sizeof(xcpt_reg_t);
>     }
> }
> ```
> 
> Or like QEMU mix ucontext and sigsetjmp/siglongjmp API.

I removed the ucontext API completely from my patch and I used the current solution with sigsetjmp/siglongjmp. I reused the setjmp/longjmp only for the initialization part (setting up the stack)

xiaoxiang781216 commented:
@patacongo sure, it's an exciting feature which can fixed many incompatible issues. @sebastianene07 please @me once you finish the updating.

sebastianene07 commented:
> @patacongo sure, it's an exciting feature which can fixed many incompatible issues. @sebastianene07 please @me once you finish the updating.

@xiaoxiang781216 The patch is ready, can you review it one more time please ? thanks, Seb

patacongo commented:
> @xiaoxiang781216 The patch is ready, can you review it one more time please ? thanks, Seb

If this is going to be merged, we should do so well before the 9.3 release which is coming up soon.

patacongo commented:
> The patch is ready, can you review it one more time please ? thanks, Seb

One thing to do that give us a lot of confidence would be to run the sim:ostest configuration.  I used to run that test like:

    $ ./nuttx 1>log 2>&1
    $ grep -i error log | grep -v nerrors=0

This will still show some strings with the word error in them, but those can be ignored.  This is a pretty good test, at least for things like this.  Round robin schedule should be enabled in the the ostest configuration to assure that pre-emption is really working.

patacongo commented:
@xiaoxiang781216 Is there any reason why this should not be merged?

xiaoxiang781216 commented:
> @xiaoxiang781216 Is there any reason why this should not be merged?

I will review and try this patch in the next couple hours and give the feedback.

sebastianene07 commented:
> > @xiaoxiang781216 Is there any reason why this should not be merged?
> 
> I will review and try this patch in the next couple hours and give the feedback.

I ran the `sim/ostest` build and I got a crash. I should further investigate it. Should I create a separate config file for this patch ?
To run the test I enabled:
`SIM_PREEMPTIBLE`, `RR_INTERVAL=1`, `SIM_WALLTIME=n`

xiaoxiang781216 commented:
Yes, it's better to add a new config to test this big feature.

sebastianene07 commented:
> Yes, it's better to add a new config to test this big feature.

LE: I fixed the issue with the crash & managed to run `sim:ostest` on both OSX & Linux. 
For the test I enabled : `SIM_PREEMPTIBLE`, `RR_INTERVAL=1` and set `SIM_WALLTIME=n`

xiaoxiang781216 commented:
@sebastianene07 I am reviewing the code now, and will provide the feedback in the weekend.

patacongo commented:
> LE: I fixed the issue with the crash & managed to run `sim:ostest` on both OSX & Linux.
> For the test I enabled : `SIM_PREEMPTIBLE`, `RR_INTERVAL=1` and set `SIM_WALLTIME=n`

Is it then ready to merge?  I would say that we set a low bar for inclusion in the repository:  The changes should not break any of the existing configurations with SIM_PREEMPTIBLE disabled.  Large new features can have bugs that take weeks to resolve and that is fine (as long as we get most of them before the 9.3 release).  So I think that if we maintain backward compatible and test a few key configurations wtih SIM_PREEMPTIBLE, we should be good to merge.

Have you tried testing the simulated SMP enabled?

sebastianene07 commented:
> > LE: I fixed the issue with the crash & managed to run `sim:ostest` on both OSX & Linux.
> > For the test I enabled : `SIM_PREEMPTIBLE`, `RR_INTERVAL=1` and set `SIM_WALLTIME=n`
> 
> Is it then ready to merge? I would say that we set a low bar for inclusion in the repository: The changes should not break any of the existing configurations with SIM_PREEMPTIBLE disabled. Large new features can have bugs that take weeks to resolve and that is fine (as long as we get most of them before the 9.3 release). So I think that if we maintain backward compatible and test a few key configurations wtih SIM_PREEMPTIBLE, we should be good to merge.
> 
> Have you tried testing the simulated SMP enabled?

The problem that I fixed was related to my patch and the current PR does not change any existing configuration (it is backwards compatible). In this form I think the patch is ready to be merged after @xiaoxiang781216 reviews it. I tested with/without `SIM_PREEMPTIBLE` option.

I checked the SMP config with `SIM_PREEMPTIBLE` disabled and it works. With `SIM_PREEMPTIBLE` enabled the SMP build doesn't seem to work because the `siglongjmp` API is not able to restore the [context created from another thread](https://pubs.opengroup.org/onlinepubs/7990989799/xsh/siglongjmp.html). The initial implementation with `ucontext` API had the advantage of supporting the SMP config but it's not portable.

patacongo commented:
> I checked the SMP config with `SIM_PREEMPTIBLE` disabled and it works. With `SIM_PREEMPTIBLE` enabled the SMP build doesn't seem to work because the `siglongjmp` API is not able to restore the [context created from another thread](https://pubs.opengroup.org/onlinepubs/7990989799/xsh/siglongjmp.html). The initial implementation with `ucontext` API had the advantage of supporting the SMP config but it's not portable.

Hmm.. I wonder if we should not retain the ucontext as an option for platforms where it is supported.  I think that ucontext support in included to Cygwin (which is my preferred Windows platform).  sys/ucontext.h exists on Cygwin and apears to be real.  I would have to experiment with that to know for certain.

The second most popular Windows environment, MSYS2, also appears to support ucontext.

sebastianene07 commented:
> > I checked the SMP config with `SIM_PREEMPTIBLE` disabled and it works. With `SIM_PREEMPTIBLE` enabled the SMP build doesn't seem to work because the `siglongjmp` API is not able to restore the [context created from another thread](https://pubs.opengroup.org/onlinepubs/7990989799/xsh/siglongjmp.html). The initial implementation with `ucontext` API had the advantage of supporting the SMP config but it's not portable.
> 
> Hmm.. I wonder if we should not retain the ucontext as an option for platforms where it is supported with something like CONFIG_SIM_HAVE_UCONTEXT. I think that ucontext support is included to Cygwin (which is my preferred Windows platform). sys/ucontext.h exists on Cygwin and apears to be real. I would have to experiment with that to know for certain.
> 
> The second most popular Windows environment, MSYS2, also appears to support `ucontext`.

From what I've tested OSX marks the `ucontext` API as deprecated and requires some special flag during compile time to use it. I think that we can provide support in a different patch for `SMP` config with `SIM_PREEMPTIBLE` option as this PR is already big. What are your thoughts on this ?

xiaoxiang781216 commented:
@sebastianene07 Does ucontext can pass ostest with SMP?

sebastianene07 commented:
> @sebastianene07 Does ucontext can pass ostest with SMP?

I didn't check but from the man pages ucontext API should support restoring the context from other threads

sebastianene07 commented:
@xiaoxiang781216 Did you have time to review my patch ? I rebased on top of the latest changes from master

xiaoxiang781216 commented:
Yes, I have reviewed the patch and tried locally, but the problem is that ostest with SMP will hang randomly. we may need to redesign the task switch code like other arch(e.g. https://github.com/apache/incubator-nuttx/blob/master/arch/arm/src/armv7-m/arm_unblocktask.c#L85):
1.Implement up_interrupt_context correctly
2.Handle host signal(either timer siganl or SMP signal) like the hardware interrupt

sebastianene07 commented:
> Yes, I have reviewed the patch and tried locally, but the problem is that ostest with SMP will hang randomly. we may need to redesign the task switch code like other arch(e.g. https://github.com/apache/incubator-nuttx/blob/master/arch/arm/src/armv7-m/arm_unblocktask.c#L85):
> 1.Implement up_interrupt_context correctly
> 2.Handle host signal(either timer siganl or SMP signal) like the hardware interrupt

The behavior you are seeing may be related to the fact that `siglongjmp` can't restore the context created from other thread, so the current implementation can't run with `SMP` config enabled.

xiaoxiang781216 commented:
Yes, sigsetjmp/siglongjmp can't be used cross the thread. As you and @patacongo investigate ucontext support on all host which we care about,  so I am thinking to implement the preemption by ucontext. But even we use ucontext API, the current implementation may still hang in SMP mode.

xiaoxiang781216 commented:
Anyway, I refine your patch more clean(only irq part), here:
https://github.com/xiaoxiang781216/incubator-nuttx/commits/preempt

sebastianene07 commented:
> Yes, sigsetjmp/siglongjmp can't be used cross the thread. As you and @patacongo investigate ucontext support on all host which we care about, so I am thinking to implement the preemption by ucontext. But even we use ucontext API, the current implementation may still hang in SMP mode.



> Anyway, I refine your patch more clean(only irq part), here:
> https://github.com/xiaoxiang781216/incubator-nuttx/commits/preempt
> We can try switch to ucontext API and see whether SMP can work correctly. If not, the context switch logic may have to redesign like other real hardware platform.

Thanks for looking on it & cleaning it up ! The current PR solves the problem with the 'fake interrupts' generated from idle task and doesn't affect other configurations if the `SIM_PREEMPTIBLE` config is off. Adding back the `ucontext API ` involves more changes like:

1. modify the way arguments are passed in a new task (we need to call `makecontext` function and provide the task arguments. I saw `up_initial_state` is called before `nxtask_setup_stackargs` so we don't have the arguments in the TCB to be able to call `makecontext` from there.)

2. Adjusting the stack size for tasks (for example on OSX the stack needs to be ~128K) 

These changes are not present in my initial version of the patch and I think is better to have a separate PR for this as it will involve big changes. What do you think ?

xiaoxiang781216 commented:
> > Yes, sigsetjmp/siglongjmp can't be used cross the thread. As you and @patacongo investigate ucontext support on all host which we care about, so I am thinking to implement the preemption by ucontext. But even we use ucontext API, the current implementation may still hang in SMP mode.
> 
> > Anyway, I refine your patch more clean(only irq part), here:
> > https://github.com/xiaoxiang781216/incubator-nuttx/commits/preempt
> > We can try switch to ucontext API and see whether SMP can work correctly. If not, the context switch logic may have to redesign like other real hardware platform.
> 
> Thanks for looking on it & cleaning it up ! The current PR solves the problem with the 'fake interrupts' generated from idle task and doesn't affect other configurations if the `SIM_PREEMPTIBLE` config is off. Adding back the `ucontext API ` involves more changes like:
> 
> 1. modify the way arguments are passed in a new task (we need to call `makecontext` function and provide the task arguments. I saw `up_initial_state` is called before `nxtask_setup_stackargs` so we don't have the arguments in the TCB to be able to call `makecontext` from there.)
> 2. Adjusting the stack size for tasks (for example on OSX the stack needs to be ~128K)
> 
> These changes are not present in my initial version of the patch and I think is better to have a separate PR for this as it will involve some big changes and it will be an improvement of my current work. What do you think ?

Since sigsetjmp/siglongjmp can't work with SMP,  we have to use ucontext API anyway. If we implement two context switch solution, this will make the code logic much complex but without benefit. I prefer to provide a solution work in all configuration.

sebastianene07 commented:
> > > Yes, sigsetjmp/siglongjmp can't be used cross the thread. As you and @patacongo investigate ucontext support on all host which we care about, so I am thinking to implement the preemption by ucontext. But even we use ucontext API, the current implementation may still hang in SMP mode.
> > 
> > 
> > > Anyway, I refine your patch more clean(only irq part), here:
> > > https://github.com/xiaoxiang781216/incubator-nuttx/commits/preempt
> > > We can try switch to ucontext API and see whether SMP can work correctly. If not, the context switch logic may have to redesign like other real hardware platform.
> > 
> > 
> > Thanks for looking on it & cleaning it up ! The current PR solves the problem with the 'fake interrupts' generated from idle task and doesn't affect other configurations if the `SIM_PREEMPTIBLE` config is off. Adding back the `ucontext API ` involves more changes like:
> > 
> > 1. modify the way arguments are passed in a new task (we need to call `makecontext` function and provide the task arguments. I saw `up_initial_state` is called before `nxtask_setup_stackargs` so we don't have the arguments in the TCB to be able to call `makecontext` from there.)
> > 2. Adjusting the stack size for tasks (for example on OSX the stack needs to be ~128K)
> > 
> > These changes are not present in my initial version of the patch and I think is better to have a separate PR for this as it will involve some big changes and it will be an improvement of my current work. What do you think ?
> 
> Since sigsetjmp/siglongjmp can't work with SMP, we have to use ucontext API anyway. If we implement two context switch solution, this will make the code logic much complex but without benefit. I prefer to provide a solution work in all configuration.

I implemented a solution with `ucontext` here : https://github.com/sebastianene07/incubator-nuttx/tree/sene/simulator_support_preemptive_scheduling_with_ucontext but it needs some cleanup. With this solution I was able to run the `sim:ostest` config with `SMP` option on and `SIM_PREEMPTIBLE`.
Some notes on this implementation:
- I implemented the handler that verifies if we are running in interrupt context by creating an alternative stack for the signals
- for all the tasks in `up_initial_state` I used the stack allocated from the host heap to prevent re-adjusting of all the stacks

sebastianene07 commented:
> > > Yes, sigsetjmp/siglongjmp can't be used cross the thread. As you and @patacongo investigate ucontext support on all host which we care about, so I am thinking to implement the preemption by ucontext. But even we use ucontext API, the current implementation may still hang in SMP mode.
> > 
> > 
> > > Anyway, I refine your patch more clean(only irq part), here:
> > > https://github.com/xiaoxiang781216/incubator-nuttx/commits/preempt
> > > We can try switch to ucontext API and see whether SMP can work correctly. If not, the context switch logic may have to redesign like other real hardware platform.
> > 
> > 
> > Thanks for looking on it & cleaning it up ! The current PR solves the problem with the 'fake interrupts' generated from idle task and doesn't affect other configurations if the `SIM_PREEMPTIBLE` config is off. Adding back the `ucontext API ` involves more changes like:
> > 
> > 1. modify the way arguments are passed in a new task (we need to call `makecontext` function and provide the task arguments. I saw `up_initial_state` is called before `nxtask_setup_stackargs` so we don't have the arguments in the TCB to be able to call `makecontext` from there.)
> > 2. Adjusting the stack size for tasks (for example on OSX the stack needs to be ~128K)
> > 
> > These changes are not present in my initial version of the patch and I think is better to have a separate PR for this as it will involve some big changes and it will be an improvement of my current work. What do you think ?
> 
> Since sigsetjmp/siglongjmp can't work with SMP, we have to use ucontext API anyway. If we implement two context switch solution, this will make the code logic much complex but without benefit. I prefer to provide a solution work in all configuration.

Hey @xiaoxiang781216, ping :) . Did you have time to view[ my new PR](https://github.com/apache/incubator-nuttx/pull/1602) with the ucontext API ?

xiaoxiang781216 commented:
I will take a look in the weekend.
